### PR TITLE
Don't copy Cargo.toml to tempdir where possible

### PR DESCRIPTION
**Pull request by illicitonion**
I'm not entirely clear on why we're copying out in the first place, but
I added a comment with my best guess.

Avoid doing this where possible, because it breaks support for Cargo
workspaces.

Alternatively, we could try to parse the Cargo.toml for its workspace
members and copy those too, but that seems to be breaking more
abstraction layers.

The limited cases of Workspaces working is probably also worth documenting, though I'm not sure where?

### CODE REVIEW

**Comment by UebelAndre** [comment]
I've wondered the same thing. I wondered if running `cargo metadata` littered your repo with some artifacts in some context. 

Loosely related, you may want to check out #193 if you're trying to do things with workspaces (just to throw in your 2 cents) 😄

**Comment by UebelAndre** [comment]
I feel like there isn't much worth doing outside of actually solving for workspace support here. Do you have any thoughts on what that would look like? I don't think it's a huge change but I could be wrong 😅

File: impl/src/metadata.rs [hunk 1]
```diff
@@ -63,9 +63,25 @@ impl MetadataFetcher for CargoMetadataFetcher {
...
     assert!(files.lock_path_opt.as_ref().map_or(true, |p| p.is_file()));
 
-    // Copy files into a temp directory
     // UNWRAP: Guarded by function assertion
-    let cargo_tempdir = {
+    let toml_dir = files.toml_path.parent().unwrap().to_path_buf();
+
+    // If the Cargo.lock file is specified and is in a different directory to the Cargo.toml,
+    // we need to copy them to a tempdir so that they're adjacent and cargo picks them up.
+    // Note that this breaks Workspace support, as the child Cargo.toml files will not be copied
```
Comment (acmcarther) [inline impl/src/metadata.rs:71]: What does "it breaks workspace support" mean?

**Review by acmcarther — COMMENTED** [review]
The code generally looks fine to me, though we probably should hold off for https://github.com/google/cargo-raze/pull/227 or one of its derived PRs will make this obsolete. Not sure though.

IIRC, the copying behavior existed so that raze woudn't modify files in the workspace. I don't know if that's valuable behavior in practice.

LMK if there's a consensus to merge this first.

**Comment by UebelAndre** [comment]
> IIRC, the copying behavior existed so that raze woudn't modify files in the workspace. I don't know if that's valuable behavior in practice.

Do you remember what files might have gotten modified?

I think it'd be great to remove the the logic for copying files into tempdirs if possible since I think that'd simplify the implementation for workspace support.

**Comment by UebelAndre** [comment]
Oh, and yeah, if this could hold off until #227 goes through that'd be helpful. I'll be as responsive as I can to make sure it gets in quickly 😄

**Comment by acmcarther** [comment]
No I've long since forgotten what it actually did.. though the git history might be illuminating. I can take a look at this in a bit.

**Comment by UebelAndre** [comment]
@acmcarther https://github.com/google/cargo-raze/pull/28 appears to be where this was introduced. Does that bring back any memories? There's nothing specifically calling out the `tempdir` stuff in that PR.

**Comment by acmcarther** [comment]
That PR is the first PR where we add an alternative to using cargo's internals to do metadata fetching.

It doesn't really clear anything up beyond what I already remembered, which is that I wanted to run the subcommand away from the actual working directory to avoid any "side effects" like generated files. This was probably more precautionary than deliberate and was probably inspired by my understanding of how bazel itself works.

**Comment by illicitonion** [comment]
https://github.com/google/cargo-raze/pull/227 added `use_tempdir` to `CargoMetadataFetcher` which fills my need here :)

**Pull request closed without merging**